ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{TirTrace, TirOp, Statement, Rvalue, Operand, Constant, ConstantInt,
+UnsignedInt};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a `u64`.
+        let func: fn() -> u64 = unsafe {
+            mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0)))
+        };
+        func()
+    }
+}
+
+/// The `TraceCompiler` takes a `SIRTrace` and compiles it to machine code. Returns a `CompiledTrace`.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    available_regs: Vec<u8>,
+    assigned_regs: HashMap<u32, u8>
+}
+
+impl TraceCompiler {
+
+    fn local_to_reg(&mut self, l: u32) -> u8 {
+        // This is a really dumb register allocator, which runs out of available registers after 7
+        // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
+        // up registers again, and allocate additional locals on the stack. Though, ultimately we
+        // probably want to implement a proper register allocator, e.g. graph colouring or linear
+        // scan.
+        match self.assigned_regs.get(&l) {
+            Some(u) => {
+                *u
+            },
+            None => {
+                if l == 0 {
+                    // Local 0 always maps to the return register RAX
+                    0
+                }
+                else {
+                    let reg = self.available_regs
+                        .pop()
+                        .expect("Can't allocate more than 7 locals yet!");
+                    self.assigned_regs.insert(l, reg);
+                    reg
+                }
+            }
+        }
+    }
+
+    pub fn local_usize(&mut self, a: u32, c: usize) {
```

Comment:
Can we add comments so that I know what `c` is here?

Reply:
Done.

Reply:
OK, that helps.

So I'm now wondering if we can come up with some better names for things.

`a` is a local index, right? So how about `l` or `local`?

The function name itself isn't super-intuitive, but I'm not sure what we'd call it?

`const_into_local`?

Reply:
Fixed in a548b5a